### PR TITLE
Bicep module to create Front Door WAF Policy

### PR DESCRIPTION
**Pull request by rajeshkaremane**
# Description

New module added to create WAF Policy for Front Door related to https://github.com/Azure/ResourceModules/issues/3714

| Pipeline |
| - |
| |

# Type of Change

- [x] New feature (non-breaking change which adds functionality)

# Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

**Review by AlexanderSehr — CHANGES_REQUESTED** [review]
Hey @rajeshkaremane, would you happen to have a successfully pipeline run handy to show that the code successfully passes all tests? For that you'd need an Azure Subscription that you set up as a target environment in your Fork.
If you don't have one available, we could also offer to rebase your PR to a dev branch on our end and test it in our environment. Please let us know. :)

File: modules/network/front-door-web-application-firewall-policy/.bicep/nested_roleAssignments.bicep [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  'Monitoring Contributor': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '749f88d5-cbae-40b8-bcfc-e573ddc772fa')
+  'Monitoring Reader': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '43d0d8ad-25c7-4714-9337-8ba259a9fe05')
+  Owner: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '8e3af657-a8ff-443c-a75c-2fe8c4bcb635')
+  Reader: subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'acdd72a7-3385-48ef-bd42-f606fba81ae7')
+  'Resource Policy Contributor': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '36243c78-bf99-498c-9df9-86d9f8d28608')
+  'Role Based Access Control Administrator (Preview)': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', 'f58310d9-a9f6-439a-9e8d-f62e7b41a168')
+  'User Access Administrator': subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '18d7d88d-d35e-4fb5-a5c3-7773c20a72d9')
+}
+
+resource fdwafpolicyProfile 'Microsoft.Network/FrontDoorWebApplicationFirewallPolicies@2022-05-01' existing = {
```
Comment (AlexanderSehr) [inline modules/network/front-door-web-application-firewall-policy/.bicep/nested_roleAssignments.bicep:50]:
```suggestion
resource frontDoorWAFPolicy 'Microsoft.Network/FrontDoorWebApplicationFirewallPolicies@2022-05-01' existing = {
```

File: modules/network/front-door-web-application-firewall-policy/.bicep/nested_roleAssignments.bicep [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  properties: {
+    description: description
+    roleDefinitionId: contains(builtInRoleNames, roleDefinitionIdOrName) ? builtInRoleNames[roleDefinitionIdOrName] : roleDefinitionIdOrName
+    principalId: principalId
+    principalType: !empty(principalType) ? any(principalType) : null
+    condition: !empty(condition) ? condition : null
+    conditionVersion: !empty(conditionVersion) && !empty(condition) ? conditionVersion : null
+    delegatedManagedIdentityResourceId: !empty(delegatedManagedIdentityResourceId) ? delegatedManagedIdentityResourceId : null
+  }
+  scope: fdwafpolicyProfile
```
Comment (AlexanderSehr) [inline modules/network/front-door-web-application-firewall-policy/.bicep/nested_roleAssignments.bicep:65]:
```suggestion
  scope: frontDoorWAFPolicy
```

File: modules/network/front-door-web-application-firewall-policy/.bicep/nested_roleAssignments.bicep [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+}]
```
Comment (AlexanderSehr) [inline modules/network/front-door-web-application-firewall-policy/.bicep/nested_roleAssignments.bicep:66]:
```suggestion
resource frontDoorWAFPolicy 'Microsoft.Network/FrontDoorWebApplicationFirewallPolicies@2022-05-01' existing = {
  name: last(split(resourceId, '/'))!
}

resource roleAssignment 'Microsoft.Authorization/roleAssignments@2022-04-01' = [for principalId in principalIds: {
  name: guid(frontDoorWAFPolicy.id, principalId, roleDefinitionIdOrName)
  properties: {
    description: description
    roleDefinitionId: contains(builtInRoleNames, roleDefinitionIdOrName) ? builtInRoleNames[roleDefinitionIdOrName] : roleDefinitionIdOrName
    principalId: principalId
    principalType: !empty(principalType) ? any(principalType) : null
    condition: !empty(condition) ? condition : null
    conditionVersion: !empty(conditionVersion) && !empty(condition) ? conditionVersion : null
    delegatedManagedIdentityResourceId: !empty(delegatedManagedIdentityResourceId) ? delegatedManagedIdentityResourceId : null
  }
  scope: frontDoorWAFPolicy
}]
```

File: .azuredevops/modulePipelines/ms.network.frontdoorwebapplicationfirewallpolicies.yml [hunk 4]
```diff
@@ -0,0 +1,50 @@
+name: 'Network - FrontDoorWebApplicationFirewallPolicies'
```
Comment (AlexanderSehr) [inline .azuredevops/modulePipelines/ms.network.frontdoorwebapplicationfirewallpolicies.yml:1]: Please also add a GitHub version of the pipeline for those in the community using them :)
You can do this by copy-pasting one of the other pipelines and just update all individual values to the one of the Front Door WAF Policies module

File: modules/network/front-door-web-application-firewall-policy/main.bicep [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  }
+  tags: tags
+  properties: {
+    customRules: customRules
+    managedRules: sku == 'Premium_AzureFrontDoor' ? managedRules : { managedRuleSets: [] }
+    policySettings: policySettings
+  }
+}
+
+resource profile_lock 'Microsoft.Authorization/locks@2020-05-01' = if (!empty(lock)) {
```
Comment (AlexanderSehr) [inline modules/network/front-door-web-application-firewall-policy/main.bicep:109]:
```suggestion
resource frontDoorWAFPolicy_lock 'Microsoft.Authorization/locks@2020-05-01' = if (!empty(lock)) {
```

File: modules/network/front-door-web-application-firewall-policy/main.bicep [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+  name: '${frontDoorWAFPolicy.name}-${lock}-lock'
+  properties: {
+    level: any(lock)
+    notes: lock == 'CanNotDelete' ? 'Cannot delete resource or child resources.' : 'Cannot modify the resource or child resources.'
+  }
+  scope: frontDoorWAFPolicy
+}
+
+module profile_roleAssignments '.bicep/nested_roleAssignments.bicep' = [for (roleAssignment, index) in roleAssignments: {
+  name: '${uniqueString(deployment().name, location)}-Profile-Rbac-${index}'
```
Comment (AlexanderSehr) [inline modules/network/front-door-web-application-firewall-policy/main.bicep:119]:
```suggestion
module frontDoorWAFPolicy_roleAssignments '.bicep/nested_roleAssignments.bicep' = [for (roleAssignment, index) in roleAssignments: {
  name: '${uniqueString(deployment().name, location)}-FDWAFP-Rbac-${index}'
```

File: modules/network/front-door-web-application-firewall-policy/.bicep/nested_roleAssignments.bicep (hunk 3 above)
Comment (rajeshkaremane) [reply modules/network/front-door-web-application-firewall-policy/.bicep/nested_roleAssignments.bicep]: Fixed in the latest commit

File: modules/network/front-door-web-application-firewall-policy/main.bicep (hunk 6 above)
Comment (rajeshkaremane) [reply modules/network/front-door-web-application-firewall-policy/main.bicep]: Fixed in the latest commit

File: modules/network/front-door-web-application-firewall-policy/main.bicep (hunk 5 above)
Comment (rajeshkaremane) [reply modules/network/front-door-web-application-firewall-policy/main.bicep]: Fixed in the latest commit

File: .azuredevops/modulePipelines/ms.network.frontdoorwebapplicationfirewallpolicies.yml (hunk 4 above)
Comment (rajeshkaremane) [reply .azuredevops/modulePipelines/ms.network.frontdoorwebapplicationfirewallpolicies.yml:1]: Added github pipeline :-)

**Comment by rajeshkaremane** [comment]
> Hey @rajeshkaremane, would you happen to have a successfully pipeline run handy to show that the code successfully passes all tests? For that you'd need an Azure Subscription that you set up as a target environment in your Fork. If you don't have one available, we could also offer to rebase your PR to a dev branch on our end and test it in our environment. Please let us know. :)

Hi @AlexanderSehr 
I am happy for you to test in dev branch.

Thanks
Rajesh

**Pull request merged**